core/src/HIP/Kokkos_HIP.cpp

```diff
@@ -66,6 +66,15 @@ void HIP::impl_initialize(InitializationSettings const& settings) {
     }
   }
 
+  // Print a warning if the user did not select the APU architecture when using
+  // a MI300A
+#ifdef KOKKOS_ARCH_AMD_GFX942
+  if ((Kokkos::show_warnings()) && (hipProp.integrated == 1)) {
+    std::cerr << "Kokkos::HIP::initialize WARNING: running kernels for MI300X "
+                 "(discrete GPU) on a MI300A (APU).\n";
+  }
+#endif
```

Comment:
Should we abort in the other direction? I.e. if you have set KOKKOS_ARCH_AMD_GFX942_APU (or for that matter if the UNIFIED_MEMORY thing is defined) but hipProp.integrated==0 we abort?

Reply:
I was thinking that if you set KOKKOS_ARCH_AMD_GFX942_APU on a MI300X, the code will crash right away and you will realize that you did something wrong. While if you set KOKKOS_ARCH_AMD_GFX942 on a MI300A, you won't get the behavior that you expect but it won't be obvious.

Reply:
Will it crash? Even if you have like the Linux memory management thingy?

Reply:
You are right. It won't crash in that case. I'll just change the warning to always warn if you pick the wrong architecture.

Reply:
If users are manually requesting `KOKKOS_ARCH_AMD_GFX942_APU` and are using `ROCm` before 6.2, they will run into a compilation error for `m_deviceProp.integrated` here since that property was only introduced in 6.2?

Reply:
It compiles fine. It just returns 0. We already have the same code in `print_configuration` and it works.